docs/rfcs/018-storage-messaging-2.md

```diff
@@ -0,0 +1,163 @@
+  positions; based on that, it decides from which safekeepers to pull WAL.
+
+Also, safekeepers use etcd elections API to make sure only single safekeeper
+offloads WAL.
+
+It works, and callmemaybe is gone. However, this has a performance
+hazard. Currently deployed etcd can do about 6k puts per second (using its own
+`benchmark` tool); on my 6 core laptop, while running on tmpfs, this gets to
+35k. Making benchmark closer to our usage [etcd watch bench](https://github.com/arssher/etcd-client/blob/watch-bench/examples/watch_bench.rs),
+I get ~10k received messages per second with various number of publisher-subscribers
+(laptop, tmpfs). Diving this by 12 (3 sks generate msg, 1 ps + 3 sk consume them) we
+get about 800 active timelines, if message is sent each second. Not extremely
+low, but quite reachable.
+
+A lot of idle watches seem to be ok though -- which is good, as pageserver
+subscribes to all its timelines regardless of their activity.
+
+Also, running etcd with fsyncs disabled is messy -- data dir must be wiped on
+each restart or there is a risk of corruption errors.
+
+The reason is etcd making much more than what we need; it is a fault tolerant
+store with strong consistency, but I claim all we need here is just simplest pub
+sub with best effort delivery, because
+* We already have centralized source of truth for long running data, like which
+  tlis are on which nodes  -- the console.
+* Momentary data (safekeeper/pageserver progress) doesn't make sense to persist.
+  Instead of putting each change to broker, expecting it to reliably deliver it
+  is better to just have constant flow of data for active timelines: 1) they
+  serve as natural heartbeats -- if node can't send, we shouldn't pull WAL from
+  it 2) it is simpler -- no need to track delivery to/from the broker.
+  Moreover, latency here is important: the faster we obtain fresh data, the
+  faster we can switch to proper safekeeper after failure.
+* As for WAL offloading leader election, it is trivial to achieve through these
+  heartbeats -- just take suitable node through deterministic rule (min node
+  id).  Once network is stable, this is a converging process (well, except
+  complicated failure topology, but even then making it converge is not
+  hard). Such elections bear some risk of several offloaders running
+  concurrently for a short period of time, but that's harmless.
+
+  Generally, if one needs strong consistency, electing leader per se is not
+  enough; it must be accompanied with number (logical clock ts), checked at
+  every action to track causality. s3 doesn't provide CAS, so it can't
+  differentiate old/new leader, this must be solved differently.
+
+  We could use etcd CAS (its most powerful/useful primitive actually) to issue
+  these leader numbers (and e.g. prefix files in s3), but currently I don't see
+  need for that.
+
+
+Obviously best effort pub sub is much more simpler and performant; the one proposed is
+
+## gRPC broker
```

Comment:
> gRPC

Can we add some paragraph with the rationale behind it?

If I understand the current model right, there has to be multiple "clients", that relatively rare per timeline (~once per second) send pulse messages  into the "dashboard" and, similarly, receive `node_number` updates per timeline in return from that dashboard.
We don't care about retries, fault tolerance, auth (?), at least now: we query the "dashboard" in a loop and get the timeline stats, otherwise considering the service broken and ignoring it.
We do not require leasing with the proposed model, emulating the lease as "node with lowest id does the job" and do not use any other features, AFAIK, just the messaging.

WIth that in mind, I feel like basic websockets are fine? Maybe even some POST request on a dashboard side + [SSE](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events) to stream the data back?

Compared to both options above, gRPC is a relatively harder protocol, `tonic` brings quite a dependency load to our project: https://github.com/hyperium/tonic/blob/master/tonic/Cargo.toml and both the protocol and its rust crates feel heavy and redundant to a stateless messaging task we're trying to achieve.

Reply:
Well, grpc provides rpc and serialization in the box, which is handy.

Do you have a particular websocket crate in mind? It might be lighter, but now sure how much. Also I guess we already have all tonic dependencies...

Reply:
> rpc and serialization in the box

_Extra_ rpc and serialization in the box, considering that pageserver has some already, and adding another http one + serde is not a churn for a "dashboard" service.

> guess we already have all tonic dependencies

Ironically, we do that because etcd requires grpc to be worked with. So if we drop etcd and stop using its client, we won't use `axium`, `tower` and probably whatever serialisation library is used there now too.
Those crates (tower especially, due to generics) are quite compilation heavy, so if we could throw all of them away, it would be nice.

> a particular websocket crate in mind

I'm afraid not, I've used some in conjunction with `actix-rt` runtime which is not our case.
I would still advocate against using anything but plain HTTP for a prototype: we don't need any new crates for that to add, SSE is naturally just a stream writing/reading and can be implemented ad-hoc.
In a simplistic model, we're talking `N_NODES` `PUT` requests with a relatively small JSON payload per timeline, with `tokio`-based simplistic http routing behind it (similar to pageserver and safekeeper ones we have), we could get good results, couldn't we?

Reply:
Heh, true.

So in nutshell, reasons for grpc were ease of programming -- out of the box streaming, rpc and serialization. And as far as I understand http2 (on which grpc stands) completely replaced websockets. But if that turns out lighter, we can try.

I'm not sure how plain post + sse is ergonomical. Also, POSTs are clearly slower than a proper stream.